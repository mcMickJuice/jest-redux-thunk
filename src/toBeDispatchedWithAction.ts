import diff from "jest-diff";
import {
  assertActionTypeProvided,
  assertIsMockFunction,
  toMatchObjectComparison
} from "./utils";
/**
 * ensures that an action was dispatched that matches a certain type
 * for actions that match that type, check each key to see if subset
 * actual dispatched thing
 *
 * similar to "toMatchObject" but with mocks
 */

interface ExpectedAction {
  type: string;
  [prop: string]: any;
}

function toBeDispatchedWithAction(
  this: jest.MatcherUtils,
  receivedDispatchMock: jest.Mock,
  expectedAction: ExpectedAction
) {
  assertIsMockFunction(receivedDispatchMock);
  assertActionTypeProvided(expectedAction.type);

  const calls = receivedDispatchMock.mock.calls;
  // check to see if dispatchMock was called at all
  if (calls.length === 0) {
    return {
      message: () => "Dispatch not called",
      pass: false
    };
  }

  const matchingActionTypeCalls = calls.filter(
    call => call[0].type === expectedAction.type
  );

  if (matchingActionTypeCalls.length === 0) {
    return {
      message: () => `No Action dispatched with type ${expectedAction.type}`,
      pass: false
    };
  }

  // TODO: support this?
  if (matchingActionTypeCalls.length > 1) {
    throw new Error(
      `Action of type ${
        expectedAction.type
      } dispatched multiple times. This matcher only supports action dispatched one time`
    );
  }

  const matchingDispatchedAction = matchingActionTypeCalls[0][0];

  const { type: _, ...otherProps } = expectedAction;

  const pass = Object.keys(otherProps).every(propKey => {
    const actual = matchingDispatchedAction[propKey];
    const expected = expectedAction[propKey];

    return toMatchObjectComparison(actual, expected);
  });

  const message = pass
    ? () =>
        `${this.utils.matcherHint(
          ".not.toBeDispatchedWithActionTypeOrder",
          "dispatch",
          "actionTypeOrder"
        )}\n\nExpected action type sequence not to have matched but it did`
    : () => {
        const diffString = diff(expectedAction, matchingDispatchedAction);

        return `Expected Dispatched Action:\n${this.utils.printExpected(
          expectedAction
        )}\n\nActual Dispatched Action Types (in order):\n${this.utils.printReceived(
          matchingDispatchedAction
        )}\n\nDifference:\n${diffString}`;
      };

  return {
    message,
    pass
  };
}

export default toBeDispatchedWithAction;
