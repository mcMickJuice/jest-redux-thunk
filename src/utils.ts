export const assertIsMockFunction = (mock: any) => {
  if (mock._isMockFunction !== true) {
    throw new Error("Value provided is not a mock function");
  }
};

export const assertActionTypeProvided = (actionType: any) => {
  if (actionType == null) {
    throw new Error("ActionType not provided");
  }
};
