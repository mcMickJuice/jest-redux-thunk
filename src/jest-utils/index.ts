import { equals } from "./jasmineUtils";
import { iterableEquality, subsetEquality } from "./utils";

export const toMatchObjectComparison = (obj1: any, obj2: any): boolean => {
  return equals(obj1, obj2, [iterableEquality, subsetEquality]);
};
