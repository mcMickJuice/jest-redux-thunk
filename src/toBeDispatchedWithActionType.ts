import { assertActionTypeProvided, assertIsMockFunction } from "./utils";

function toBeDispatchedWithActionType(
  this: jest.MatcherUtils,
  receivedDispatchMock: jest.Mock,
  expectedActionType: string
) {
  assertIsMockFunction(receivedDispatchMock);
  assertActionTypeProvided(expectedActionType);

  const calls = receivedDispatchMock.mock.calls;
  // check to see if dispatchMock was called at all
  if (calls.length === 0) {
    return {
      message: () => "Dispatch not called",
      pass: false
    };
  }

  // assume always called with one param
  const pass = calls.some(call => call[0].type === expectedActionType);

  const message = pass
    ? () =>
        `${this.utils.matcherHint(
          ".not.dispatchedWithActionType",
          "dispatch",
          "actionType"
        )} \n\n Expected action type ${this.utils.printExpected(
          expectedActionType
        )} not to have been dispatched but it was`
    : () =>
        `${this.utils.matcherHint(
          ".dispatchedWithActionType",
          "dispatch",
          "actionType"
        )} \n\n Expected action type ${this.utils.printExpected(
          expectedActionType
        )} to have been dispatched but it wasn't`;

  return {
    message,
    pass
  };
}

export default toBeDispatchedWithActionType;
