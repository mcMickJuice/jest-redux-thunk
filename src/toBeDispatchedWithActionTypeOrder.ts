import diff from "jest-diff";
import deepEqual from "lodash.isequal";
import { assertIsMockFunction } from "./utils";

function toBeDispatchedWithActionTypeOrder(
  this: jest.MatcherUtils,
  dispatchMock: jest.Mock,
  expectedActionTypeOrder: string[]
) {
  assertIsMockFunction(dispatchMock);

  const calls = dispatchMock.mock.calls;
  const actualDispatchedActionTypes = calls.map(call => call[0].type);

  const pass = deepEqual(expectedActionTypeOrder, actualDispatchedActionTypes);

  const message = pass
    ? () =>
        `${this.utils.matcherHint(
          ".not.toBeDispatchedWithActionTypeOrder",
          "dispatch",
          "actionTypeOrder"
        )}\n\nExpected action type sequence not to have matched but it did`
    : () => {
        const diffString = diff(
          expectedActionTypeOrder,
          actualDispatchedActionTypes
        );

        return `Expected Dispatched Action Types (in order):\n${this.utils.printExpected(
          expectedActionTypeOrder
        )}\n\nActual Dispatched Action Types (in order):\n${this.utils.printReceived(
          actualDispatchedActionTypes
        )}\n\nDifference:\n${diffString}`;
      };

  return {
    message,
    pass
  };
}

export default toBeDispatchedWithActionTypeOrder;
