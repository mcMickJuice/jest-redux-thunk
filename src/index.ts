import toBeDispatchedWithAction from "./toBeDispatchedWithAction";
import toBeDispatchedWithActionType from "./toBeDispatchedWithActionType";
import toBeDispatchedWithActionTypeOrder from "./toBeDispatchedWithActionTypeOrder";

expect.extend({
  toBeDispatchedWithAction,
  toBeDispatchedWithActionType,
  toBeDispatchedWithActionTypeOrder
});
