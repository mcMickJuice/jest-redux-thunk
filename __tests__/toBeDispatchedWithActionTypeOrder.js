require("../src");

describe("toBeDispatchedWithActionTypeOrder", () => {
  const actionType1 = "ACTION_TYPE_1";
  const actionType2 = "ACTION_TYPE_2";
  const actionType3 = "ACTION_TYPE_3";
  const expectedActionTypeOrder = [actionType1, actionType2, actionType3];
  describe("invalid parameters", () => {
    it("fails if mock function not passed to expect", () => {
      const myFunc = function fn() {
        /** */
      };
      let error;
      try {
        expect(myFunc).toBeDispatchedWithActionTypeOrder(
          expectedActionTypeOrder
        );
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
  });

  describe("affirmative tests", () => {
    it("passes if actionTypes are dispatched in correct order", () => {
      const dispatchMock = jest.fn();

      dispatchMock({
        type: actionType1
      });

      dispatchMock({
        type: actionType2
      });
      dispatchMock({
        type: actionType3
      });

      expect(dispatchMock).toBeDispatchedWithActionTypeOrder(
        expectedActionTypeOrder
      );
    });

    it("fails if actionTypes are not dispatched in correct order", () => {
      const dispatchMock = jest.fn();

      dispatchMock({
        type: actionType1
      });

      dispatchMock({
        type: actionType3
      });

      dispatchMock({
        type: actionType2
      });

      let error;

      try {
        expect(dispatchMock).toBeDispatchedWithActionTypeOrder(
          expectedActionTypeOrder
        );
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });

    it("fails if one or more expect action types is not dispatched", () => {
      const dispatchMock = jest.fn();

      dispatchMock({
        type: actionType1
      });

      dispatchMock({
        type: actionType2
      });

      let error;

      try {
        expect(dispatchMock).toBeDispatchedWithActionTypeOrder(
          expectedActionTypeOrder
        );
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });

    it("fails if no dispatches are made", () => {
      const dispatchMock = jest.fn();

      let error;

      try {
        expect(dispatchMock).toBeDispatchedWithActionTypeOrder(
          expectedActionTypeOrder
        );
      } catch (e) {
        error = e;
      }

      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
  });

  describe("negative tests", () => {
    it("passes if action type dispatch order doesnt match", () => {
      const dispatchMock = jest.fn();

      dispatchMock({
        type: actionType1
      });

      dispatchMock({
        type: actionType2
      });

      expect(dispatchMock).not.toBeDispatchedWithActionTypeOrder([
        actionType2,
        actionType1
      ]);
    });
    it("fails if action type dispatch order matches", () => {
      const dispatchMock = jest.fn();

      dispatchMock({
        type: actionType1
      });

      dispatchMock({
        type: actionType2
      });
      dispatchMock({
        type: actionType3
      });

      let error;
      try {
        expect(dispatchMock).not.toBeDispatchedWithActionTypeOrder(
          expectedActionTypeOrder
        );
      } catch (e) {
        error = e;
      }
      expect(error).toBeDefined();
      expect(error.message).toMatchSnapshot();
    });
  });
});
