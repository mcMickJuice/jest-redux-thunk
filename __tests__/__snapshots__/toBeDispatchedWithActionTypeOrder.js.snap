// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeDispatchedWithActionTypeOrder affirmative tests fails if actionTypes are not dispatched in correct order 1`] = `
"Expected Dispatched Action Types (in order):
[32m[\\"ACTION_TYPE_1\\", \\"ACTION_TYPE_2\\", \\"ACTION_TYPE_3\\"][39m

Actual Dispatched Action Types (in order):
[31m[\\"ACTION_TYPE_1\\", \\"ACTION_TYPE_3\\", \\"ACTION_TYPE_2\\"][39m

Difference:
[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[2m    \\"ACTION_TYPE_1\\",[22m
[31m+   \\"ACTION_TYPE_3\\",[39m
[2m    \\"ACTION_TYPE_2\\",[22m
[32m-   \\"ACTION_TYPE_3\\",[39m
[2m  ][22m"
`;

exports[`toBeDispatchedWithActionTypeOrder affirmative tests fails if no dispatches are made 1`] = `
"Expected Dispatched Action Types (in order):
[32m[\\"ACTION_TYPE_1\\", \\"ACTION_TYPE_2\\", \\"ACTION_TYPE_3\\"][39m

Actual Dispatched Action Types (in order):
[31m[][39m

Difference:
[32m- Expected[39m
[31m+ Received[39m

[32m- Array [[39m
[32m-   \\"ACTION_TYPE_1\\",[39m
[32m-   \\"ACTION_TYPE_2\\",[39m
[32m-   \\"ACTION_TYPE_3\\",[39m
[32m- ][39m
[31m+ Array [][39m"
`;

exports[`toBeDispatchedWithActionTypeOrder affirmative tests fails if one or more expect action types is not dispatched 1`] = `
"Expected Dispatched Action Types (in order):
[32m[\\"ACTION_TYPE_1\\", \\"ACTION_TYPE_2\\", \\"ACTION_TYPE_3\\"][39m

Actual Dispatched Action Types (in order):
[31m[\\"ACTION_TYPE_1\\", \\"ACTION_TYPE_2\\"][39m

Difference:
[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[2m    \\"ACTION_TYPE_1\\",[22m
[2m    \\"ACTION_TYPE_2\\",[22m
[32m-   \\"ACTION_TYPE_3\\",[39m
[2m  ][22m"
`;

exports[`toBeDispatchedWithActionTypeOrder invalid parameters fails if mock function not passed to expect 1`] = `"Value provided is not a mock function"`;

exports[`toBeDispatchedWithActionTypeOrder negative tests fails if action type dispatch order matches 1`] = `
"[2mexpect([22m[31mdispatch[39m[2m).not.toBeDispatchedWithActionTypeOrder([22m[32mactionTypeOrder[39m[2m)[22m

Expected action type sequence not to have matched but it did"
`;
